Tensile/SolutionStructs.py

```diff
@@ -44,12 +46,277 @@ def pvar(state, field):
 def roundupRatio(dividend, divisor):
   return int(math.ceil(float(dividend) / float(divisor)))
 
+class DimAB(Enum):
+  OnlyA = 0  # Only allowed in A tensor
+  OnlyB = 1  # Only allowed in B tensor
+  BothAB = 2 # Must be in both A and B
+
+class Fbs:
+  Free=0     # Expect to be free dimension
+  Batch=1    # Expect to be batch dimension
+  Sum=2      # Expect to be summation dimension
+
+################################################################################
+class Convolution:
+  class Dimension:
+    # stride=-1 indicates TBD stride; >=0 indicates a compile-time constant
+    def __init__(self, shortChar, description, dimAB, strideA=-1, strideB=-1):
+      self.shortChar = shortChar
+      self.description = description
+      self.dimAB = dimAB
+      self.strideA=strideA
+      self.strideB=strideB
+
+    def __str__(self):
+      return "%5s : %s" % ("'%s'"%self.shortChar, self.description)
+    def __repr__(self):
+      return self.shortChar
+
+  def dimxParm(self, config, parmName, default):
+    parm =config.get(parmName)
+    if not parm:
+      rv = [default ] * self.formatNumSpatialDims
+    else:
+      rv=[]
+      for x in parm.split('x'):
+        if x.upper()=='N':
+          rv.append(-1)
+        else:
+          rv.append(int(x))
+      rv.reverse() # rightmost number is 0
+    if len(rv) != self.formatNumSpatialDims:
+        raise RuntimeError ("%s parm '%s' must have %d spatial dims'"%(parmName, parm, self.formatNumSpatialDims))
+    return rv
+
+  def printUsage(self, problemType):
+    print("Tensile Index Assignments and Usage:")
+    print("Tensile  : ConvChar: Explanation/Usage")
+    for (idx,dim) in enumerate(self.indexAssignments):
+      if dim != None:
+        tensileChar = globalParameters['IndexChars'][idx]
+        usage = str(dim)
+        usage = usage.replace('#T#', tensileChar)
+        usage = usage.replace('#S0#', str(self.stride[0]))
+        usage = usage.replace('#S1#', str(self.stride[1]))
+        usage = usage.replace('#D0#', str(self.dilation[0]))
+        print("  %d('%c') :   %s" % (idx, tensileChar, usage))
+    print ()
+    print ("- '(TBD)' indicates the parm is flexible and must be specified at runtime.")
+    print ("- '(i)' where i is an integer constant, indicates the parm is hard-coded at compile time.")
+    print ("  The runtime value must match the compile-time value.")
+    print ("- Unspecified strides use default stride value:")
+    print ("    stride[i] = (stride[i-1]*size[i]) for i>0 ; 1 for i==0")
+
+    print ()
+    print ("ProblemType Definition:")
+    for k in (
+        'OperationType','DestDataType','DataType','HighPrecisionAccumulate',\
```

Comment:
This list could be separated out into a class property.

Reply:
can you say more - the list of problem type properties could be a @property of Convolution class?

Reply:
The goal is to make it more obvious why these specific properties are relevant here and to make it possible to use this list elsewhere.

I don't specifically mean using `@property` - I guess class member or class variable would be more accurate: just assign these to a variable at the class level, say `PropertyNames` or `PropertyKeys`.  Then you would access it here with` Convolution.PropertyNames`.